### ISSUE TITLE
eval: incorrect error message when working with list comprehensions

### ISSUE THREAD

**Issue by nichtsundniemand** [cue-lang/cue#3007]
### What version of CUE are you using (`cue version`)?

<pre>
$ cue version
cue version v0.8.1

go version go1.22.0
      -buildmode exe
       -compiler gc
  DefaultGODEBUG httplaxcontentlength=1,httpmuxgo121=1,tls10server=1,tlsrsakex=1,tlsunsafeekm=1
     CGO_ENABLED 1
          GOARCH amd64
            GOOS linux
         GOAMD64 v1
</pre>

### Does this issue reproduce with the latest stable release?
As shown above I'm using CUE v0.8.1, so yes.

### What did you do?
I was working with a list defined by a mix of comprehension and direct element declarations, like this:
```
l: [1, 2, 3] & [
    for e in a {e},
    b, c,
]
a: [1, 2]
b: 3
c: 4
```

This should, of course, not work, and it doesn't. To my surprise though the error-message is a bit confusing.

### What did you expect to see?
A sensible error-message.

### What did you see instead?
This:
```
l: incompatible list lengths (3 and 3)
```

### Additional info
This appears to work properly:
```
l: [
    for e in a {e},
    b, c,
]
a: [1, 2]
b: 3
c: 4
l: [1, 2, 3]
```
Giving this message, which is what I would expect:
```
l: incompatible list lengths (3 and 4)
```

OTOH this also fails with the message implying 3!=3:
```
l: [1, 2, 3]
l: [
    for e in a {e},
    b, c,
]
a: [1, 2]
b: 3
c: 4
```

So it appears like the comprehension is sometimes counted as a single element and sometimes counted as all elements produced by the comprehension.

I have some more examples that may help clarify the issue attached in this file:
[miscount.cue.txt](https://github.com/cue-lang/cue/files/14863806/miscount.cue.txt)

**Comment by tvandinther** [comment]
I just encountered this as well. Small repro: https://cuelang.org/play/?id=mfJpqVbW1Ya#w=function&i=cue&f=eval&o=cue

```cue
x: [int]
x: [for v in [1, 2, 3] { v }]
```
produces
```
x: incompatible list lengths (1 and 1)
```

**Comment by myitcv** [comment]
Confirmed with 5d2da079a295:

### What did you do?

```
# preamble for evalv3
env CUE_EXPERIMENT=evalv3
env CUE_DEBUG=openinline=0

# verify error message
! exec cue export x.cue
stderr 'incompatible list lengths \(1 and 3\)'

-- x.cue --
x: [int]
x: [for v in [1, 2, 3] {v}]
```

### What did you expect?

Passing test.

### What did you see instead?

```
# preamble for evalv3 (0.000s)
# verify error message (0.050s)
> ! exec cue export x.cue
[stderr]
x: incompatible list lengths (1 and 1):
    ./x.cue:2:4
[exit status 1]
> stderr 'incompatible list lengths \(1 and 3\)'
FAIL: /tmp/testscript4025072527/repro.txtar/script.txtar:7: no match for `incompatible list lengths \(1 and 3\)` found in stderr
```

**Comment by nichtsundniemand** [comment]
I've been looking into this abit further, trying to pin down the cause. While writing I re-ran everything against 9e333c606194edc184ffd518be71509a4b52bc1c.

TL;DR: I don't think it's feasible to make the error message show the correct list lengths in this case because comprehensions are evaluated somewhat lazily.
If we don't want an incorrect message, I propose we just show something like `comprehension exceeds list length ($max_length)`.

The message itself is produced by [`internal/core/adt.nodeContext.invalidListLength()`](https://github.com/cue-lang/cue/blob/master/internal/core/adt/eval.go#L2624). Throwing a `debug.PrintStack()` in there shows the relevant callsite of that method to be inside [`nodeContext.addLists()`](https://github.com/cue-lang/cue/blob/master/internal/core/adt/eval.go#L2531):
```go
// Terminate early in case of runaway comprehension.
if !isOpen && int(index) > max {
	n.invalidListLength(max, len(l.list.Elems), maxNode, l.list)
	continue outer
}
```

This makes sense since the comprehensions in our examples exceed the maximum allowed list length. I assume that's what was meant by "runaway comprehension".

Looking at `addLists()` further also confirms my initial suspicion that "the comprehension is sometimes counted as a single element and sometimes counted as all elements produced by the comprehension". The method [loops over all list elements](https://github.com/cue-lang/cue/blob/master/internal/core/adt/eval.go#L2490), when an element of type `Comprehension` is encountered, the comprehension is [evaluated and expanded by a call to `c.yield()`](https://github.com/cue-lang/cue/blob/master/internal/core/adt/eval.go#L2493).

The closure passed to yield then does the counting of elements produced by the comprehension only as a side-effect of [incrementing the current list index](https://github.com/cue-lang/cue/blob/master/internal/core/adt/eval.go#L2496) for each item yielded.

After processing each list element the current index is compared to the max length to detect a runaway comprehension, as shown above.

Since the values passed to `invalidListLength()` are `max` (in the reduced example the length of `[int]`, just a single element) and `len(l.list.Elems)` (int the reduced example the length of `[for v in [...] {v}]`, just a single comprehension) we get the confusing message with "1 and 1".

Incidentally, I think I understand now why reversing the order of comprehension and closed list (see "Additional info" in the opening post) produces a correct error message: In this case the roles are reversed. In `max` will be the actual, evaluated length of the comprehension since that was "seen" first and evaluation hasn't been constrained. So we don't actually encounter a "runaway comprehension" but just a usual case of incompatible list lengths.

In an attempt to fix the problem i tried replacing `len(l.list.Elems)` with just `index` since that should contain the current real size of the list:
```diff
diff --git a/internal/core/adt/eval.go b/internal/core/adt/eval.go
index 800faf2b..9e74c90a 100644
--- a/internal/core/adt/eval.go
+++ b/internal/core/adt/eval.go
@@ -2528,7 +2528,7 @@ outer:
 
 			// Terminate early in case of runaway comprehension.
 			if !isOpen && int(index) > max {
-				n.invalidListLength(max, len(l.list.Elems), maxNode, l.list)
+				n.invalidListLength(max, int(index), maxNode, l.list)
 				continue outer
 			}
 		}
```

Running that against the reduced example indeed produces the correct version of the error message:
```
x: incompatible list lengths (1 and 3)
```

Unfortunately this fails some test:
* [Unification of two closed lists of different length using go-API](https://github.com/cue-lang/cue/blob/master/cue/context_test.go#L63):
  ```
  --- FAIL: TestNewList (0.00s)
      --- FAIL: TestNewList/#05 (0.00s)
          context_test.go:85:  got: _|_ // incompatible list lengths (0 and 1)
              want: _|_ // incompatible list lengths (0 and 3)
  ```
* [Unification of two closed lists of different length directly in CUE](https://github.com/cue-lang/cue/blob/master/cmd/cue/cmd/testdata/script/exp_gengotypes.txtar#L138):
  ```
  -fail.cue."18_IntListClosed2".types.IntListClosed2: incompatible list lengths (2 and 3)
  +fail.cue."18_IntListClosed2".types.IntListClosed2: incompatible list lengths (2 and 4)
  ```

This makes sense since `addLists()` throws the error at the first element exceeding the maximum length which is not necessarily also the last element of the offending list. *BUT*: I believe the comment about runaway comprehensions may be outdated since that `if` clause obviously also catches length mismatches that aren't caused by any comprehension.

My second (crude) attempt at fixing the issue looked like this:
```diff
diff --git a/internal/core/adt/eval.go b/internal/core/adt/eval.go
index 800faf2b..a74b937c 100644
--- a/internal/core/adt/eval.go
+++ b/internal/core/adt/eval.go
@@ -2488,12 +2488,14 @@ outer:
 		index := int64(0)
 		hasComprehension := false
 		for j, elem := range l.list.Elems {
+			var compLen int64
 			switch x := elem.(type) {
 			case *Comprehension:
 				err := c.yield(nil, l.env, x, state, func(e *Environment) {
 					label, err := MakeLabel(x.Source(), index, IntLabel)
 					n.addErr(err)
 					index++
+					compLen++
 					c := MakeConjunct(e, x.Value, l.id)
 					n.insertField(label, ArcMember, c)
 				})
@@ -2528,7 +2530,11 @@ outer:
 
 			// Terminate early in case of runaway comprehension.
 			if !isOpen && int(index) > max {
-				n.invalidListLength(max, len(l.list.Elems), maxNode, l.list)
+				elems := int64(len(l.list.Elems))
+				if hasComprehension {
+					elems = elems - 1 + compLen
+				}
+				n.invalidListLength(max, int(elems), maxNode, l.list)
 				continue outer
 			}
 		}
```
Basically: Keep track of the actual element count produced by the comprehension and add that to the total when printing the error message.

This produces the correct error message for the reduced example and passes all unit tests. But we can easily construct another example that breaks it again:
```cue
package test

a: [0, 1, 2]
b: [3, 4, 5]
c: [int, int, int, int]
c: [for e in a {e}, for e in b {e}]
```
And we get:
```
c: incompatible list lengths (4 and 4)
```

This happens because my fix only tracks the total for the _current_ comprehension, and not for any previous ones. After evaluating the first comprehension the maximum is still not exceeded, so the second one is evaluated. That causes the length to overflow pushing execution into the error-branch. The error-message logic takes the length of the current comprehension (3) and adds it to the number of previous elements in the list (which is 1, a single comprehension) yielding 4 instead of six.

This may be fixed by counting the elements produced by all comprehensions seens so far. But a similar example shows a flaw with that approach since the evaluator will not (and probably shouldn't) look ahead:
```cue
package test

a: [0, 1, 2]
b: [3, 4, 5]
c: [int, int]
c: [for e in a {e}, for e in b {e}]
```
The error message produced is:
```
c: incompatible list lengths (2 and 4)
```
Here the second comprehension is never evaluated so there is no way to count the elements produced by it.

AFAICT there is no way to statically get the number of elements produced by a comprehension. The `Yielder` interface has no way to get the number of elements. A simple for-loop doing a 1-to-1 mapping like in our examples may be easy to solve since `len(input) == len(output)` but any comprehension that is not bijective (for example conditionally skipping elements of the input or producing more than one output value per input value) is more difficult if not impossible.

Because of this I think the only way to get the correct number of elements in an offending list that contains multiple comprehensions is to evaluate all comprehensions and tally them up. That seems overkill for an error message and directly subverts the runaway-check causing the issue in the first place...

**Comment by myitcv** [comment]
Confirming that as of c479844b8d7a (which is now in tip) this problem remains, but also that we do not need to set `CUE_DEBUG=openinline=0` in order to reproduce. i.e. we see the same behaviour with:

* `CUE_DEBUG` unset
* `CUE_DEBUG=openinline=0`
* `CUE_DEBUG=openinline=1`
